lib/lf/geometry/geometry_interface.h

```diff
@@ -125,6 +125,43 @@ class Geometry {
    */
   virtual std::unique_ptr<Geometry> SubGeometry(dim_t codim, dim_t i) const = 0;
 
+  /**
+   * @brief Generate a geometry object arising in the course of refinement
+   * 
+   * @param ref_pattern A code indicating the particular refinement of the element
+   * @paream selector A number selecting a particular child element
+   *
+   * Implementation of local mesh refinement entails splitting of elements into
+   * parts ("children"), whose shape will depend on the refinement pattern. 
+   * This method creates the geometry objects describing the shape of children.
+   *
+   * The arguments are unspecified integers to be interpreted correctly by
+   * specializations of this method. At the time of the definition of the interface
+   * all possible refinement patterns cannot be predicted. This is why  
+   * vanilla integer arguments have been chosen for this method.
+   */
+  virtual std::unique_ptr<Geometry>
+  ChildGeometry(int ref_pattern,int anchor,int selector) const {
```

Comment:
I have to say that I strongly disagree with the signature of this method. As you write it in the comment, not all possible refinement patterns can be predicted but I think using an integer argument to solve this problem is exactly the wrong approach:
- First of all, using integers is not very explanatory, if at all, I would use an enum and extend the enum every time a new type of refinement is needed.
- The much more important problem with this approach is the following: So far, we have only 4 different classes that implement the geometry interface: `Point`, `Segment`, `TriaO1` and `QuadO1`. I think in the future there are going to be many more implementations for second-order, maybe third-order elements (of Lagrangian and maybe Chebyshev type). Also, there could be implementations of NURBS surfaces and implementations that employ Transfinite Interpolation to construct the cell geometry...  Now imagine what happens if you decide to add only one more refinement pattern: You will have to change all the classes that implement the Geometry interface.

I think there are far better alternatives to using an integer argument to specify the refinement pattern. We can start by assuming that every refinement pattern should split a geometry object into smaller geometry objects that do not intersect with each other and that together constitute the whole (father) geometry object (except in cases where the father geometry merely approximates the real boundary). Also, it makes sense to define the refinement pattern on the reference element. I think a simple solution that would work for 95% of the cases is to describe the refinement pattern by a class:

```
class RefinementPattern {
 public:
  /** Reference element that is refined */
  base::RefEl RefEl();

  /**
   * Coordinate of the points that make up this refinement pattern (must all
   * lie inside the RefEl()).
   * By Convetion `PointCoord.col(i) == RefEl().NodeCoord().col(i)` for
   * `i < RefEl.NumPoints()`.
   */
  Eigen::VectorXd PointCoords();

  /** Total number of cells into which the reference element is split */
  unsigned int NumCells();

  /**
   * Get the zero-based indices of the points of the `i`-th cell.
   */
  std::vector<unsigned int> CellPoints(unsigned int i);

  /**
   * @brief The Reference Element of the i-th cell.
   */
  base::RefEl CellRefEl(unsigned int i);
};
```

The `ChildGeometry()` method could then just accept a `RefinementPattern` object. As you can see the RefinementPattern class is a simplified description of a Mesh but with the following restrictions:
- It describes implicitly a first-order mesh, i.e. the father geometry can only be split into first order sub-elements (on the reference element). Note that this does not pose a problem for cases where the father geometry only approximates the real boundary as in this case the refinement pattern is still defined on the reference element, but the mapping of the child geometries is adapted accordingly.
- It does not give the full set of topology information that is present in a normal mesh, but this information can be constructed if necessary (similar to how it is done in the constructor of the Mesh classes).

I think it is clear that the above implementation works rather well in cases where the father geometry object represents the exact geometry. As we have already discussed in issue #6 the situation is a little bit more complicated if the father geometry only approximates the real (surface) geometry. In this case, the ChildGeometry Method must somehow have a parametrization of the relevant surfaces and there are in principle two ways to achieve this:
1.  Add an additional parameter to the `ChildGeometry()` method
2.  Inject the correct parametrization into the geometry object when it is constructed.

Currently I'm not 100% sure about what is the better solution, but what is important is that there are ways that the `ChildGeometry()` method knows about the real shape of the surface(s). Depending on the type of the father geometry object, the `ChildGeometry()` should then produce the child geometries that interpolate the surface better. We can study the case of a first-order Lagrangian Triangle element: In this case the `ChildGeometry()` algorithm would have to do roughly the following:
1. Determine which points of the `RefinementPattern` lie on surfaces that are only approximated. 
2. Determine the global coordinates of these points by using the exact representation of the bounding surfaces. The global coordinates of all other points could be determined by using the geometry of the father element.
3. Construct the child triangle geometry objects with the new point coordinates.

The process will be very similar for higher order lagrangian/chebyshev geometry implementations.
With my proposal of the `RefinementPattern` Step 1 will involve some distance checks in order to find out if a node lies on a particular surface. I know that this is not perfect, but this could be resolved by adding more member functions to the `RefinementPattern` class that give the necessary information.

Reply:
Concerning cells with "higher-order" edge geometry I am planning to endow the cell `Geometry` itself with that information. Therefore `ChildGeometry()` wil work  without passing additional information.

Reply:
I believe it is important to put derived classes in the `Geometry` hierarchy in charge of the handling of geometries in the course of refinement. This will make it possible to implement new versions of `Geometry` objects without tampering with the refinement code at all.

Reply:
I am going to create a lf/refine directory and put all refinement related code there. I will also add one more level to the `Geometry` class tree: `lf::refine::RefTriaO1` etc. These classes will provide nothing else by the `ChildGeometry()` method. I hope  to have finished the bulk of the implementation of refinement by the end of July 2018.

It is important to allow the implementation of `Geonetry` classes that (initially) do not support refinement.